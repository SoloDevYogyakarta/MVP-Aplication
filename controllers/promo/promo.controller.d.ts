import { Response } from 'express';
import { CustomRequest } from '../../types/custom-request.type';
import { ProductPromoRepository } from '../../repository/promo/promo.repository';
import { DyanmicQuery } from '../../validators/query/product.query';
import { PromoField } from '../../validators/promo/promo.validator';
import { ProductPromoService } from '../../services/promo/promo.service';
export declare class PromoController {
    private readonly repository;
    private readonly service;
    constructor(repository: ProductPromoRepository, service: ProductPromoService);
    created(req: CustomRequest, res: Response, body: PromoField): Promise<Response<any, Record<string, any>>>;
    updated(req: CustomRequest, res: Response, body: PromoField): Promise<Response<any, Record<string, any>>>;
    list(query: DyanmicQuery, req: CustomRequest, res: Response): Promise<Response<any, Record<string, any>>>;
    detail(req: CustomRequest, res: Response): Promise<Response<any, Record<string, any>>>;
    destroy(req: CustomRequest, res: Response): Promise<Response<any, Record<string, any>>>;
}
