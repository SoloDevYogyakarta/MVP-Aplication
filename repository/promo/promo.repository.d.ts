import { DyanmicQuery } from '../../validators/query/product.query';
import { ProductPromoInstance } from '../../database/entities/products/promo-entity/promo-entity';
export declare class ProductPromoRepository {
    findOne(public_id: string): Promise<ProductPromoInstance>;
    findAll(query: DyanmicQuery): Promise<ProductPromoInstance[]>;
}
