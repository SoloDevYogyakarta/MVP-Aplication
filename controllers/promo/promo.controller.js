"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromoController = void 0;
const common_1 = require("@nestjs/common");
const promo_repository_1 = require("../../repository/promo/promo.repository");
const product_query_1 = require("../../validators/query/product.query");
const promo_validator_1 = require("../../validators/promo/promo.validator");
const promo_service_1 = require("../../services/promo/promo.service");
const guards_middleware_1 = require("../../middleware/guards.middleware");
let PromoController = class PromoController {
    constructor(repository, service) {
        this.repository = repository;
        this.service = service;
    }
    async created(req, res, body) {
        const result = await this.service.create(body, req.user.data.public_id);
        return res.status(result.status).json(result);
    }
    async updated(req, res, body) {
        const result = await this.service.update(body, req.params.id, req.user.data.public_id);
        return res.status(result.status).json(result);
    }
    async list(query, req, res) {
        const result = this.repository.findAll(query);
        return res.status(common_1.HttpStatus.OK).json(result);
    }
    async detail(req, res) {
        const result = await this.repository.findOne(req.params.id);
        return res.status(common_1.HttpStatus.OK).json(result);
    }
    async destroy(req, res) {
        const result = await this.service.destroy(req.params.id, req.user.data.public_id);
        return res.status(result.status).json(result);
    }
};
exports.PromoController = PromoController;
__decorate([
    (0, common_1.Post)(),
    (0, common_1.UseGuards)(guards_middleware_1.AuthGuard),
    __param(0, (0, common_1.Req)()),
    __param(1, (0, common_1.Res)()),
    __param(2, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, promo_validator_1.PromoField]),
    __metadata("design:returntype", Promise)
], PromoController.prototype, "created", null);
__decorate([
    (0, common_1.Post)(':id'),
    (0, common_1.UseGuards)(guards_middleware_1.AuthGuard),
    __param(0, (0, common_1.Req)()),
    __param(1, (0, common_1.Res)()),
    __param(2, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, promo_validator_1.PromoField]),
    __metadata("design:returntype", Promise)
], PromoController.prototype, "updated", null);
__decorate([
    (0, common_1.Get)(),
    __param(0, (0, common_1.Query)()),
    __param(1, (0, common_1.Req)()),
    __param(2, (0, common_1.Res)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [product_query_1.DyanmicQuery, Object, Object]),
    __metadata("design:returntype", Promise)
], PromoController.prototype, "list", null);
__decorate([
    (0, common_1.Get)(':id'),
    __param(0, (0, common_1.Req)()),
    __param(1, (0, common_1.Res)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], PromoController.prototype, "detail", null);
__decorate([
    (0, common_1.Delete)(':id'),
    (0, common_1.UseGuards)(guards_middleware_1.AuthGuard),
    __param(0, (0, common_1.Req)()),
    __param(1, (0, common_1.Res)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], PromoController.prototype, "destroy", null);
exports.PromoController = PromoController = __decorate([
    (0, common_1.Controller)('promo'),
    __metadata("design:paramtypes", [promo_repository_1.ProductPromoRepository,
        promo_service_1.ProductPromoService])
], PromoController);
//# sourceMappingURL=promo.controller.js.map