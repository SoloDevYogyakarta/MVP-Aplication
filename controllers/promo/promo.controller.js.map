{"version":3,"file":"promo.controller.js","sourceRoot":"","sources":["../../../src/controllers/promo/promo.controller.ts"],"names":[],"mappings":";;;;;;;;;;;;;;;AACA,2CAWwB;AAExB,8EAAiF;AACjF,wEAAoE;AACpE,4EAAoE;AACpE,sEAAyE;AACzE,0EAA+D;AAGxD,IAAM,eAAe,GAArB,MAAM,eAAe;IAC1B,YACmB,UAAkC,EAClC,OAA4B;QAD5B,eAAU,GAAV,UAAU,CAAwB;QAClC,YAAO,GAAP,OAAO,CAAqB;IAC5C,CAAC;IAIE,AAAN,KAAK,CAAC,OAAO,CACJ,GAAkB,EAClB,GAAa,EACZ,IAAgB;QAExB,MAAM,MAAM,GAAG,MAAM,IAAI,CAAC,OAAO,CAAC,MAAM,CAAC,IAAI,EAAE,GAAG,CAAC,IAAI,CAAC,IAAI,CAAC,SAAS,CAAC,CAAC;QACxE,OAAO,GAAG,CAAC,MAAM,CAAC,MAAM,CAAC,MAAM,CAAC,CAAC,IAAI,CAAC,MAAM,CAAC,CAAC;IAChD,CAAC;IAIK,AAAN,KAAK,CAAC,OAAO,CACJ,GAAkB,EAClB,GAAa,EACZ,IAAgB;QAExB,MAAM,MAAM,GAAG,MAAM,IAAI,CAAC,OAAO,CAAC,MAAM,CACtC,IAAI,EACJ,GAAG,CAAC,MAAM,CAAC,EAAE,EACb,GAAG,CAAC,IAAI,CAAC,IAAI,CAAC,SAAS,CACxB,CAAC;QACF,OAAO,GAAG,CAAC,MAAM,CAAC,MAAM,CAAC,MAAM,CAAC,CAAC,IAAI,CAAC,MAAM,CAAC,CAAC;IAChD,CAAC;IAGK,AAAN,KAAK,CAAC,IAAI,CACC,KAAmB,EACrB,GAAkB,EAClB,GAAa;QAEpB,MAAM,MAAM,GAAG,IAAI,CAAC,UAAU,CAAC,OAAO,CAAC,KAAK,CAAC,CAAC;QAC9C,OAAO,GAAG,CAAC,MAAM,CAAC,mBAAU,CAAC,EAAE,CAAC,CAAC,IAAI,CAAC,MAAM,CAAC,CAAC;IAChD,CAAC;IAGK,AAAN,KAAK,CAAC,MAAM,CAAQ,GAAkB,EAAS,GAAa;QAC1D,MAAM,MAAM,GAAG,MAAM,IAAI,CAAC,UAAU,CAAC,OAAO,CAAC,GAAG,CAAC,MAAM,CAAC,EAAE,CAAC,CAAC;QAC5D,OAAO,GAAG,CAAC,MAAM,CAAC,mBAAU,CAAC,EAAE,CAAC,CAAC,IAAI,CAAC,MAAM,CAAC,CAAC;IAChD,CAAC;IAIK,AAAN,KAAK,CAAC,OAAO,CAAQ,GAAkB,EAAS,GAAa;QAC3D,MAAM,MAAM,GAAG,MAAM,IAAI,CAAC,OAAO,CAAC,OAAO,CACvC,GAAG,CAAC,MAAM,CAAC,EAAE,EACb,GAAG,CAAC,IAAI,CAAC,IAAI,CAAC,SAAS,CACxB,CAAC;QACF,OAAO,GAAG,CAAC,MAAM,CAAC,MAAM,CAAC,MAAM,CAAC,CAAC,IAAI,CAAC,MAAM,CAAC,CAAC;IAChD,CAAC;CACF,CAAA;AAzDY,0CAAe;AAQpB;IAFL,IAAA,aAAI,GAAE;IACN,IAAA,kBAAS,EAAC,6BAAS,CAAC;IAElB,WAAA,IAAA,YAAG,GAAE,CAAA;IACL,WAAA,IAAA,YAAG,GAAE,CAAA;IACL,WAAA,IAAA,aAAI,GAAE,CAAA;;qDAAO,4BAAU;;8CAIzB;AAIK;IAFL,IAAA,aAAI,EAAC,KAAK,CAAC;IACX,IAAA,kBAAS,EAAC,6BAAS,CAAC;IAElB,WAAA,IAAA,YAAG,GAAE,CAAA;IACL,WAAA,IAAA,YAAG,GAAE,CAAA;IACL,WAAA,IAAA,aAAI,GAAE,CAAA;;qDAAO,4BAAU;;8CAQzB;AAGK;IADL,IAAA,YAAG,GAAE;IAEH,WAAA,IAAA,cAAK,GAAE,CAAA;IACP,WAAA,IAAA,YAAG,GAAE,CAAA;IACL,WAAA,IAAA,YAAG,GAAE,CAAA;;qCAFU,4BAAY;;2CAM7B;AAGK;IADL,IAAA,YAAG,EAAC,KAAK,CAAC;IACG,WAAA,IAAA,YAAG,GAAE,CAAA;IAAsB,WAAA,IAAA,YAAG,GAAE,CAAA;;;;6CAG7C;AAIK;IAFL,IAAA,eAAM,EAAC,KAAK,CAAC;IACb,IAAA,kBAAS,EAAC,6BAAS,CAAC;IACN,WAAA,IAAA,YAAG,GAAE,CAAA;IAAsB,WAAA,IAAA,YAAG,GAAE,CAAA;;;;8CAM9C;0BAxDU,eAAe;IAD3B,IAAA,mBAAU,EAAC,OAAO,CAAC;qCAGa,yCAAsB;QACzB,mCAAmB;GAHpC,eAAe,CAyD3B"}