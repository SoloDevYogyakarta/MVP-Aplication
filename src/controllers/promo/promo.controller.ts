import { Response } from 'express';
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpStatus,
  Post,
  Query,
  Req,
  Res,
  UseGuards,
} from '@nestjs/common';
import { CustomRequest } from '../../types/custom-request.type';
import { ProductPromoRepository } from '../../repository/promo/promo.repository';
import { DyanmicQuery } from '../../validators/query/product.query';
import { PromoField } from '../../validators/promo/promo.validator';
import { ProductPromoService } from '../../services/promo/promo.service';
import { AuthGuard } from '../../middleware/guards.middleware';

@Controller('promo')
export class PromoController {
  constructor(
    private readonly repository: ProductPromoRepository,
    private readonly service: ProductPromoService,
  ) {}

  @Post()
  @UseGuards(AuthGuard)
  async created(
    @Req() req: CustomRequest,
    @Res() res: Response,
    @Body() body: PromoField,
  ) {
    const result = await this.service.create(body, req.user.data.public_id);
    return res.status(result.status).json(result);
  }

  @Post(':id')
  @UseGuards(AuthGuard)
  async updated(
    @Req() req: CustomRequest,
    @Res() res: Response,
    @Body() body: PromoField,
  ) {
    const result = await this.service.update(
      body,
      req.params.id,
      req.user.data.public_id,
    );
    return res.status(result.status).json(result);
  }

  @Get()
  async list(
    @Query() query: DyanmicQuery,
    @Req() req: CustomRequest,
    @Res() res: Response,
  ) {
    const result = await this.repository.findAll(query);
    return res.status(HttpStatus.OK).json(result);
  }

  @Get(':id')
  async detail(@Req() req: CustomRequest, @Res() res: Response) {
    const result = await this.repository.findOne(req.params.id);
    return res.status(HttpStatus.OK).json(result);
  }

  @Delete(':id')
  @UseGuards(AuthGuard)
  async destroy(@Req() req: CustomRequest, @Res() res: Response) {
    const result = await this.service.destroy(
      req.params.id,
      req.user.data.public_id,
    );
    return res.status(result.status).json(result);
  }
}
