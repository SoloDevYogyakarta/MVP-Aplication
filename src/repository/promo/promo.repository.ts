import { Injectable } from '@nestjs/common';
import { dynamicFilter } from '../../utils/dynamic-filter/dynamic-filter';
import { DyanmicQuery } from '../../validators/query/product.query';
import {
  productPromoAssciate,
  productPromoAttribute,
  productPromoInclude,
} from '../../database/associates/promo-associate/promo-associate';
import { ProductPromoInstance } from '../../database/entities/products/promo-entity/promo-entity';

@Injectable()
export class ProductPromoRepository {
  async findOne(public_id: string): Promise<ProductPromoInstance> {
    return await productPromoAssciate.findOne({
      where: { public_id },
      attributes: productPromoAttribute,
      include: productPromoInclude,
    });
  }

  async findAll(query: DyanmicQuery): Promise<ProductPromoInstance[]> {
    let where = {},
      datas = [];
    where = dynamicFilter(where, datas, query);
    return await productPromoAssciate.findAll({
      where: where,
      attributes: productPromoAttribute,
      include: productPromoInclude,
    });
  }
}
