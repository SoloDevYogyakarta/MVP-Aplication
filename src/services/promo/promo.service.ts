import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { userEntity } from '../../database/entities/authenticates/user-entity/user-entity';
import { productBasicEntity } from '../../database/entities/products/basic-entity/basic-entity';
import { PromoField } from '../../validators/promo/promo.validator';
import { omit } from 'lodash';
import { productPromoEntity } from '../../database/entities/products/promo-entity/promo-entity';
import { createpath } from '../../utils/system/system';

@Injectable()
export class ProductPromoService {
  async create(body: PromoField, user_id: string) {
    await this.isAdmin(user_id);
    const findOne = await productBasicEntity.findOne({
      where: { public_id: body.product_id },
    });
    if (!findOne) {
      throw new HttpException(
        {
          status: HttpStatus.NOT_FOUND,
          message: 'Product not found',
        },
        HttpStatus.NOT_FOUND,
      );
    }
    const create = await productPromoEntity.create(body);
    create.save();
    createpath('../../database/dataTxt/promo-http-entity.txt', create);
    return { status: HttpStatus.CREATED, message: 'Promo has been create' };
  }

  async update(body: PromoField, public_id: string, user_id: string) {
    await this.isAdmin(user_id);
    const findOne = await productPromoEntity.findOne({ where: { public_id } });
    if (!findOne) {
      throw new HttpException(
        {
          status: HttpStatus.NOT_FOUND,
          message: 'Promo not found',
        },
        HttpStatus.NOT_FOUND,
      );
    }
    findOne.update(omit(body, ['product_id']), { where: { public_id } });
    return { status: HttpStatus.OK, message: 'Promo has been update' };
  }

  async destroy(public_id: string, user_id: string) {
    await this.isAdmin(user_id);
    const findOne = await productPromoEntity.findOne({ where: { public_id } });
    if (!findOne) {
      throw new HttpException(
        { status: HttpStatus.NOT_FOUND, message: 'Promo not found' },
        HttpStatus.NOT_FOUND,
      );
    }
    findOne.destroy();
    return { status: HttpStatus.OK, message: 'Promo has been delete' };
  }

  async isAdmin(public_id: string): Promise<void> {
    const findOne = await userEntity.findOne({ where: { public_id } });
    if (findOne?.role !== 'admin') {
      throw new HttpException('false', HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }
}
