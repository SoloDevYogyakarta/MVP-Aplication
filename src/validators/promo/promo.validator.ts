import { IsDate, IsNumber, IsOptional, IsString } from 'class-validator';

export class PromoField {
  @IsNumber()
  main_stock!: number;

  @IsNumber()
  reverse_stock!: number;

  @IsNumber()
  value!: number;

  @IsDate()
  start_time!: Date;

  @IsDate()
  end_time!: Date;

  @IsString()
  @IsOptional()
  product_id?: string;
}
