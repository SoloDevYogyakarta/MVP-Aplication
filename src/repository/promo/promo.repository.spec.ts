import { ProductPromoRepository } from './promo.repository';

describe('ProductPromoRepository', () => {
  let public_id!: string;
  let repository!: ProductPromoRepository;

  beforeEach(() => {
    repository = new ProductPromoRepository();
  });

  it('should to be defined', () => expect(repository).toBeDefined());

  it('render correctly', () => expect(repository).toMatchSnapshot());

  if (public_id) {
    it('findAll', async () => {
      const result = await repository.findAll({});
      expect(result).not.toEqual(0);
    });

    it('findOne', async () => {
      const result = await repository.findOne(public_id);
      expect(result.public_id).toEqual(public_id);
    });
  }
});
