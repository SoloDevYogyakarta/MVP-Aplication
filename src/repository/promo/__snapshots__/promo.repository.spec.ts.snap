// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductPromoRepository render correctly 1`] = `ProductPromoRepository {}`;
